### PR TITLE
pref: Implement structured logging

### PR DESCRIPTION
**Pull request by engedaam**
Fixes https://github.com/aws/karpenter/issues/1465

**Description**
- Implemented structured Logging
**How was this change tested?**

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/deprovisioning/emptiness.go [hunk 1]
```diff
@@ -55,7 +55,10 @@ func (e *Emptiness) ShouldDeprovision(ctx context.Context, n *state.Node, provis
 
 	emptinessTime, err := time.Parse(time.RFC3339, emptinessTimestamp)
 	if err != nil {
-		logging.FromContext(ctx).Debugf("Unable to parse emptiness timestamp, %s for node %s", emptinessTimestamp, n.Node.Name)
+		ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With(
+			"potimestampd", emptinessTimestamp,
```
Comment (jonathan-innis) [inline pkg/controllers/deprovisioning/emptiness.go:59]:
```suggestion
			"emptinessTimestamp", emptinessTimestamp,
```

File: pkg/controllers/deprovisioning/emptiness.go [hunk 2]
```diff
@@ -55,7 +55,10 @@ func (e *Emptiness) ShouldDeprovision(ctx context.Context, n *state.Node, provis
...
+			"nodeName", n.Node.Name))
```
Comment (jonathan-innis) [inline pkg/controllers/deprovisioning/emptiness.go:60]:
```suggestion
			"node", n.Node.Name))
```

File: pkg/controllers/deprovisioning/expiration.go [hunk 3]
```diff
@@ -81,7 +81,8 @@ func (e *Expiration) ComputeCommand(ctx context.Context, attempt int, candidates
 	}
 	// Log when all pods can't schedule, as the command will get executed immediately.
 	if !allPodsScheduled {
-		logging.FromContext(ctx).Infof("continuing to expire node %s after scheduling simulation failed to schedule all pods", node.Name)
+		ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("nodeName", node.Name))
```
Comment (jonathan-innis) [inline pkg/controllers/deprovisioning/expiration.go:84]:
```suggestion
		ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("node", node.Name))
```

Comment (jonathan-innis) [inline pkg/controllers/deprovisioning/expiration.go:84]: We should probably inject the node name after retrieving it on line 72, that way, if there are other log statements added later, they can also use the name

File: pkg/controllers/deprovisioning/emptiness.go [hunk 4]
```diff
@@ -55,7 +55,10 @@ func (e *Emptiness) ShouldDeprovision(ctx context.Context, n *state.Node, provis
 
 	emptinessTime, err := time.Parse(time.RFC3339, emptinessTimestamp)
```
Comment (jonathan-innis) [inline pkg/controllers/deprovisioning/emptiness.go:56]: Musing: since the node is always passed into this interface method, can we always pass the node name into the logger encapsulated by the caller in the `controller.go` file?

File: pkg/controllers/provisioning/provisioner.go [hunk 5]
```diff
@@ -111,7 +111,8 @@ func (p *Provisioner) Builder(_ context.Context, mgr manager.Manager) operatorco
 			// wait to ensure that our cluster state is synced with the current known nodes to prevent over-provisioning
 			for WaitForClusterSync {
 				if err := p.cluster.Synchronized(ctx); err != nil {
-					logging.FromContext(ctx).Infof("waiting for cluster state to catch up, %s", err)
+					ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("clusterState", err))
```
Comment (jonathan-innis) [inline pkg/controllers/provisioning/provisioner.go:114]:
```suggestion
```

Errors don't really work well as structured values since they aren't consistent and hard to parse

File: pkg/controllers/provisioning/provisioner.go [hunk 6]
```diff
@@ -328,7 +329,7 @@ func (p *Provisioner) launch(ctx context.Context, opts LaunchOptions, node *sche
 		return cheapestOfferingPrice(iOfferings, node.Requirements) < cheapestOfferingPrice(jOfferings, node.Requirements)
 	})
 
-	logging.FromContext(ctx).Infof("Launching %s", node)
+	logging.FromContext(ctx).With("node", node).Infof("Launching Node")
```
Comment (jonathan-innis) [inline pkg/controllers/provisioning/provisioner.go:332]: Can we inject this at the top of this function since this function is scoped to the node?

File: pkg/controllers/provisioning/provisioner.go [hunk 7]
```diff
@@ -350,7 +351,9 @@ func (p *Provisioner) launch(ctx context.Context, opts LaunchOptions, node *sche
 	// before the node is fully Ready.
 	if _, err := p.coreV1Client.Nodes().Create(ctx, k8sNode, metav1.CreateOptions{}); err != nil {
 		if errors.IsAlreadyExists(err) {
-			logging.FromContext(ctx).Debugf("node %s already registered", k8sNode.Name)
+			ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With(
+				"nodeName", k8sNode.Name))
```
Comment (jonathan-innis) [inline pkg/controllers/provisioning/provisioner.go:355]:
```suggestion
				"node", k8sNode.Name))
```

File: pkg/controllers/provisioning/provisioner.go (hunk 6 above)
Comment (jonathan-innis) [reply pkg/controllers/provisioning/provisioner.go:332]: What does this output? I haven't looked at the `String()` method for the `*scheduler.Node` but it might be too long to be considered a structured value

Comment (jonathan-innis) [inline pkg/controllers/provisioning/provisioner.go:332]: Inject this at the top of the func?

File: pkg/controllers/provisioning/scheduling/scheduler.go [hunk 8]
```diff
@@ -163,7 +163,7 @@ func (s *Scheduler) recordSchedulingResults(ctx context.Context, pods []*v1.Pod,
 	if newCount == 0 {
 		return
 	}
-	logging.FromContext(ctx).Infof("Found %d provisionable pod(s)", len(pods))
+	logging.FromContext(ctx).With("numberOfPods", len(pods)).Infof("Found provisionable pod(s)")
```
Comment (jonathan-innis) [inline pkg/controllers/provisioning/scheduling/scheduler.go:166]:
```suggestion
	logging.FromContext(ctx).With("podCount", len(pods)).Infof("Found provisionable pod(s)")
```

**Review by jonathan-innis — CHANGES_REQUESTED** [review]
Overall, good work! 🎉 Only overarching feedback is that we should inject values the earliest we possibly can. If there is a function that is scoped to a particular thing (like operations on a node), we can inject at the top of this func, rather than somewhere in the middle of it

File: pkg/controllers/provisioning/provisioner.go (hunk 6 above)
Comment (engedaam) [reply pkg/controllers/provisioning/provisioner.go:332]: Yeah, it's better as a string format

File: pkg/controllers/deprovisioning/emptiness.go (hunk 4 above)
Comment (engedaam) [reply pkg/controllers/deprovisioning/emptiness.go:56]: Sounds good! Updated the value to be in the `controller.go` file

File: pkg/controllers/deprovisioning/emptiness.go [hunk 9]
```diff
@@ -55,7 +55,7 @@ func (e *Emptiness) ShouldDeprovision(ctx context.Context, n *state.Node, provis
...
 	if err != nil {
-		logging.FromContext(ctx).Debugf("Unable to parse emptiness timestamp, %s for node %s", emptinessTimestamp, n.Node.Name)
+		logging.FromContext(ctx).With("emptinessTimestamp", emptinessTimestamp).Debugf("Unable to parse emptiness timestamp")
```
Comment (ellistarn) [inline pkg/controllers/deprovisioning/emptiness.go:58]:
```suggestion
		logging.FromContext(ctx).With("emptiness-timestamp", emptinessTimestamp).Debugf("Unable to parse emptiness timestamp")
```

File: pkg/controllers/provisioning/scheduling/preferences.go [hunk 10]
```diff
@@ -49,8 +49,7 @@ func (p *Preferences) Relax(ctx context.Context, pod *v1.Pod) bool {
 		if reason := relaxFunc(pod); reason != nil {
 
 			logging.FromContext(ctx).
-				With("pod", client.ObjectKeyFromObject(pod)).
-				Debugf("Relaxing soft constraints for pod since it previously failed to schedule, %s", ptr.StringValue(reason))
+				With("pod", client.ObjectKeyFromObject(pod)).Debugf("Relaxing soft constraints for pod since it previously failed to schedule, %s", ptr.StringValue(reason))
```
Comment (jonathan-innis) [inline pkg/controllers/provisioning/scheduling/preferences.go:52]: Can this be moved to the top of the function?

**Review by jonathan-innis — COMMENTED** [review]
One comment

**Review by ellistarn — APPROVED** [review]
Nice work!
